src/buffer/out/textBuffer.cpp

```diff
@@ -2533,17 +2559,36 @@ void TextBuffer::_AppendRTFText(std::string& contentBuilder, const std::wstring_
     }
 }
 
-void TextBuffer::Serialize(const wchar_t* destination) const
+void TextBuffer::SerializeToPath(const wchar_t* destination) const
 {
     const wil::unique_handle file{ CreateFileW(destination, GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_DELETE, nullptr, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, nullptr) };
     THROW_LAST_ERROR_IF(!file);
 
     static constexpr size_t writeThreshold = 32 * 1024;
+    ChunkedSerialize(writeThreshold, L"\uFEFF", nullptr, [&](std::wstring& buffer, bool /* isDone */) {
+        const auto fileSize = gsl::narrow<DWORD>(buffer.size() * sizeof(wchar_t));
+        DWORD bytesWritten = 0;
+        THROW_IF_WIN32_BOOL_FALSE(WriteFile(file.get(), buffer.data(), fileSize, &bytesWritten, nullptr));
+        THROW_WIN32_IF_MSG(ERROR_WRITE_FAULT, bytesWritten != fileSize, "failed to write");
+        buffer.clear();
+    });
+}
+
+// Routine Description:
+// - Serializes the text buffer in chunks including ANSI escape code control sequences, and passes those chunks to the callback once they're large enough.
+// Arguments:
+// - writeThreshold - How big the chunk needs to be before passing it to the `chunkReadyCallback`. If 0, won't pass to the `chunkReadyCallback` until fully done serializing (effectively turns off the chunking mechanism).
+// - prefix - A string to put at the beginning of the very first chunk.
+// - copyReq - If present, controls the bounds of the selected region and other related configuration flags. If not present, the whole TextBuffer is serialized.
+// - chunkReadyCallback - Called with the chunk when the chunk is bigger than the `writeThreshold` (if non-zero) and at the very end of the serialization process (in which case `isDone` will be `true`).
+void TextBuffer::ChunkedSerialize(const size_t writeThreshold, const std::wstring prefix, const CopyRequest* copyReq, std::function<void(std::wstring& chunk, const bool isDone)> chunkReadyCallback) const

```

Comment:
I wasn't sure about using a pointer for `copyReq`. Some internet searching turned up a possibility of a `std::optional<std::reference_wrapper<CopyRequest>>` but I wasn't sure if that's better or worse.

Reply:
Since `CopyRequest` is trivial data, you can create a copy and save it inside `std::optional` without having a huge penalty. So, `std::optional<CopyRequest>` should work in this case.

Reply:
Good news is this is no longer relevant, but that makes sense for the future! 🙏

Reply:
`std::optional<std::reference_wrapper<CopyRequest>>`

It may be worth mentioning that C++ does not have a nullable optimization like Rust does. In other words, a optional reference type like that is not pointer-sized and not a lightweight, trivial type. Additionally, the MSVC x64 ABI that we need to use does not support passing structs larger than a pointer by value. In other words, when you write
```cpp
struct Foo {
    int a;
};
void foo(optional<reference_wrapper<Foo>> f);
```
what you get is a pointer to a pointer to a value:
```cpp
struct Foo {
    int a;
};
struct ReferenceWrapper {
    Foo* ptr;
};
struct Optional {
    ReferenceWrapper value;
    bool exists;
};
void foo(Optional* f);
```
That's why I think raw pointers are the perfect nullable reference. They're just fine.